### PR TITLE
contextual_lenses project folder

### PR DESCRIPTION
**Pull request by amirshane**
fixed comments from previous pull request

### CODE REVIEW

File: contextual_lenses/test_contextual_lenses/test_pooling.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+         test3,
+         test4,
+         test5,
+         test6,
+         test7,
+         test8)
+
+
+class TestPooling(parameterized.TestCase):
+  """Abstract method for testing mean and max pool reduce functions."""
```
Comment (davidBelanger) [inline contextual_lenses/test_contextual_lenses/test_pooling.py:113]: inaccurate docstring. I'd just say """Tests for mean and max pool reduce functions."""

File: contextual_lenses/train_utils.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    return loss
+  
+  grad_fn = jax.value_and_grad(compute_loss_fn)
+  _, grad = grad_fn(optimizer.target, X, Y, loss_fn, loss_fn_kwargs)
+  optimizer = optimizer.apply_gradient(grad)
+  
+  return optimizer
+
+
+def train(model, train_data, loss_fn, loss_fn_kwargs, learning_rate=1e-4, weight_decay=0.1,
```
Comment (davidBelanger) [inline contextual_lenses/train_utils.py:47]: you should have a test somewhere that if you write state to disk and load back in you get the same thing back. Especially important when restoring optimizers.

File: contextual_lenses/train_utils.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  if save_dir is not None:
+    checkpoints.save_checkpoint(ckpt_dir=save_dir, target=optimizer, step=optimizer.state.step)
+
+  return optimizer
+
+
+p_train_step = jax.pmap(train_step, axis_name='batch', static_broadcasted_argnums=(3, 4))
+
+
+def p_train(model, train_data, loss_fn, loss_fn_kwargs, learning_rate=1e-4, weight_decay=0.1,
```
Comment (davidBelanger) [inline contextual_lenses/train_utils.py:69]: lots of code duplication. Can you just have a flag called 'use_pmap' for train()?

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
+"""Contextual lenses on Pfam family classification dataset (first 100 families)."""
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py:1]: It would be good if we had a working doc for summarizing these results. I think the easiest thing would be to use google slides.

Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py:1]: It would be good if we had a working doc for summarizing these results. I think the easiest thing would be to use google slides.

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+train_df['mod_family_accession'] = train_df.family_accession.apply(lambda x: mod_family_accession(x))
+test_df['mod_family_accession'] = test_df.family_accession.apply(lambda x: mod_family_accession(x))
+dev_df['mod_family_accession'] = dev_df.family_accession.apply(lambda x: mod_family_accession(x))
+
+kept_family_accessions = []
+for i in range(1, 101):
+  kept_family_accessions.append('PF00' + '0'*(3-len(str(i))) + str(i))
+
+def keep_family_accessions(df, kept_family_accessions):
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py:83]: you can replace this  function with df[df.mod_family_accession.isin(kept_family_accessions)]

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+train_df['one_hot_inds'] = train_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+test_df['one_hot_inds'] = test_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+dev_df['one_hot_inds'] = dev_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+
+print(train_df.head())
+
+
+"""Data batching."""
+def create_data_iterator(df, batch_size, epochs=1, buffer_size=None, seed=0, drop_remainder=False):
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py:133]: as you said earlier, this should be merged with the create_data_iterator function for gfp. Just input df and the name of the columns to use for inputs and outputs. Necessary for the PR, but good to do in medium term.

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+del cnn_max_pool_model
+
+cnn_max_pool_results, cnn_max_pool_preds = evaluate(predict_fn=cnn_max_pool_optimizer.target,
+                                                    test_data=test_batches,
+                                                    true_indexes=test_indexes,
+                                                    title='CNN + Max Pool',
+                                                    loss_fn_kwargs=loss_fn_kwargs)
+
+print(cnn_max_pool_results)
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py:223]: instead of writing a proper unit test for the model saving/loading, at the very least you could do something at the bottom of this file where you load a new model from disk, compute the eval metrics, and make sure they match cnn_max_pool_results

File: contextual_lenses/tape_gfp/tape_gfp_one_hot_lenses_parallel.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
+"""Contextual lenses on TAPE fluoresence dataset."""
```
Comment (davidBelanger) [inline contextual_lenses/tape_gfp/tape_gfp_one_hot_lenses_parallel.py:1]: As above, I would look to remove this file and just have one for tape_gfp. You can have a flag for whether to pmap or not.

File: contextual_lenses/train_utils.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    X, Y = batch
+    optimizer = train_step(optimizer, X, Y, loss_fn, loss_fn_kwargs)
+  
+  if save_dir is not None:
+    checkpoints.save_checkpoint(ckpt_dir=save_dir, target=optimizer, step=optimizer.state.step)
+
+  return optimizer
+
+
+p_train_step = jax.pmap(train_step, axis_name='batch', static_broadcasted_argnums=(3, 4))
```
Comment (davidBelanger) [inline contextual_lenses/train_utils.py:66]: don't make this a global variable, as doing so will mean that jax.pmap will be called every time you import the module.
Instead wrap this in a function like get_p_train_step() or something.

File: contextual_lenses/contextual_lenses.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+from operator import itemgetter
+
+
+def max_pool(x, padding_mask=None):
+  """Apply padding, take maximum over sequence length axis."""
+
+  if padding_mask is not None:
+    x = x * padding_mask
+    neg_mask = - 999999. * (1 - padding_mask)
```
Comment (dmrd) [inline contextual_lenses/contextual_lenses.py:24]: make this a default argument to function. Set to something like -1e8 by default

File: contextual_lenses/contextual_lenses.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    rep = jnp.sum(x, axis=-2) / jnp.sum(padding_mask, axis=-2)
+  else:
+    rep = jnp.mean(x, axis=-2)
+
+  return rep
+
+
+def linear_max_pool(x, rep_size, padding_mask=None):
+  """Apply linear transformation + ReLU, apply padding,
+     take maximum over sequence length."""
```
Comment (dmrd) [inline contextual_lenses/contextual_lenses.py:46]: style nit: Put the ending """ on its own line. Same with comment at top of file.

May be worth running a python formatter over if you don't have one already.  Such as https://github.com/psf/black

File: contextual_lenses/encoders.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+  """Applies one-hot encoding followed by 1D CNN."""
+
+  one_hots = one_hot_encoder(batch_inds, num_categories)
+  cnn_one_hots = CNN(one_hots, n_layers, n_features, n_kernel_sizes)
+  
+  return cnn_one_hots
+
+
+"""Positional embeddings
+Code source: https://github.com/google/flax/blob/aff10f032e892e28a1acf4dd4ee9dcc6cd39a606/examples/wmt/models.py."""
```
Comment (dmrd) [inline contextual_lenses/encoders.py:55]: style: use # to add comments in middle of file. Generally only use """ for docstrings and the one at top of files.

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
+"""Contextual lenses on Pfam family classification dataset (first 100 families)."""
+
+
+"""Parse command line arguments."""
+from parser import parse_args
+tpu_name, save_dir, restore_dir = parse_args()
+
+
+"""Connect TPU to VM instance."""
```
Comment (dmrd) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py:9]: style nit (here and elsewhere): single line comments

# Connect TPU to VM instance.

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+  
+  for batch in iter(test_data):
+
+    X, Y = batch
+
+    Y_hat = predict_fn(X)
+
+    cross_entropy += cross_entropy_loss(Y, Y_hat, **loss_fn_kwargs)
+
+    preds = jnp.argmax(jax.nn.log_softmax(Y_hat), axis=1)
```
Comment (dmrd) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py:163]: the softmax is unnecessary here: it maintains the relative orders, so the argmax can take just Y_hat

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py (hunk 4 above)
Comment (dmrd) [reply contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py:1]: +1 slides for results. Nice way to keep a research log, and has side effect of being immediately presentable :)
